KEYCLOAK-13047 LDAP no-import fixes. Avoid lost updates - dont allow …

mposolda opened the pull request:
…update attributes, which are not mapped to LDAP

mposolda committed:
KEYCLOAK-13047 LDAP no-import fixes. Avoid lost updates - dont allow update attributes, which are not mapped to LDAP

mposolda commented:
Codebase PR: https://github.com/keycloak/keycloak/pull/7038

iankko commented on server_admin/topics/user-federation/ldap.adoc:
```diff
@@ -17,12 +17,12 @@ Selecting _ldap_ will bring you to the LDAP configuration page.
 By default, {project_name} will import users from LDAP into the local {project_name} user database. This copy of the user
 is either synchronized on demand, or through a periodic background task.
-The one exception to this is passwords.  Passwords are not imported and password validation is
+The one exception to this is passwords.  Passwords are not imported and password validation is always
```
Nitpicks:
```
s/The one exception to this is passwords./The single exception to this is the synchronization of passwords./
s/Passwords are not imported and password validation is always .../Passwords are never imported. Their validation is always delegated to the LDAP server./
```

mposolda replied:
I will change as you suggested.

iankko commented on server_admin/topics/user-federation/ldap.adoc:
```diff
@@ -17,12 +17,12 @@ Selecting _ldap_ will bring you to the LDAP configuration page.
-from the {project_name} database the 2nd time they are accessed.  The only load your LDAP server will have is password validation.
-The downside to this approach is that when a user is first queried, this will require a {project_name} database insert.  The import will
-also have to be synchronized with your LDAP server as needed.
+per-user data that is needed can be stored locally.
```
```s/The benefits to this approach.../The benefits of this approach.../g```

mposolda replied:
Changing as you suggested

iankko commented on server_admin/topics/user-federation/ldap.adoc:
```diff
@@ -17,12 +17,12 @@ Selecting _ldap_ will bring you to the LDAP configuration page.
-The downside to this approach is that when a user is first queried, this will require a {project_name} database insert.  The import will
-also have to be synchronized with your LDAP server as needed.
+per-user data that is needed can be stored locally.
+The downside to this approach is that when a user is first queried, this will require a {project_name} database insert. The import may
```
```s/The downside to this approach is that when a user is first queried.../The downside of this approach being each time a specific user is queried for the first time, a corresponding {project_name} database insert is performed./```

mposolda replied:
Will change as you suggested.

iankko commented on server_admin/topics/user-federation/ldap.adoc:
```diff
@@ -17,12 +17,12 @@ Selecting _ldap_ will bring you to the LDAP configuration page.
+per-user data that is needed can be stored locally.
+The downside to this approach is that when a user is first queried, this will require a {project_name} database insert. The import may
+also have to be synchronized with your LDAP server as needed. However import synchronization is not necessary in many cases as you can
+configure LDAP mappers in a way, that particular attributes of the user are preferably always read from the LDAP rather than from the DB.
```
```s/However import synchronization is not necessary in many cases.../However, import synchronization is not necessary in the case the LDAP mappers are configured to always read particular attributes from the LDAP rather than from the DB./```

mposolda replied:
Will change as you suggested.

iankko commented on server_admin/topics/user-federation/ldap.adoc:
```diff
@@ -32,6 +32,13 @@ The benefit to this approach is that you do not have the overhead of importing a
 
 This storage mode is controled by the `Import Users` switch.  Set to `On` to import users.
 
+NOTE: In case that import of users is disabled, there is no support to save user profile attributes and most of other user metadata
```
```s/In case/In the case/g```

mposolda replied:
Will change as you suggested.

iankko commented on server_admin/topics/user-federation/ldap.adoc:
```diff
@@ -32,6 +32,13 @@ The benefit to this approach is that you do not have the overhead of importing a
 This storage mode is controled by the `Import Users` switch.  Set to `On` to import users.
 
+NOTE: In case that import of users is disabled, there is no support to save user profile attributes and most of other user metadata
+      into the {project_name} database. The only exception are user profile metadata, which are mapped to the LDAP. This possibly includes
```
```s/The only exception/The single exception to this/g```

mposolda replied:
Changing as you suggested

iankko commented on server_admin/topics/user-federation/ldap.adoc:
```diff
@@ -32,6 +32,13 @@ The benefit to this approach is that you do not have the overhead of importing a
+NOTE: In case that import of users is disabled, there is no support to save user profile attributes and most of other user metadata
+      into the {project_name} database. The only exception are user profile metadata, which are mapped to the LDAP. This possibly includes
+      role mappings, group mappings and other metadata based on the configuration of your LDAP mappers.
+      When the attempt is made to change some non-LDAP mapped user data, the update of the user will not be possible. For example
```
```s/to change some non-LDAP mapped/some of the non-LDAP mapped/g```

mposolda replied:
I believe you meant:
`s/to change some non-LDAP mapped/to change some of the non-LDAP mapped/g` ?

I changed it that way.

iankko replied:
Yes, it was just about adding `of` word (sorry for typo in the copy-pasting). You got it right.

iankko commented on server_admin/topics/user-federation/ldap.adoc:
```diff
@@ -51,6 +58,13 @@ UNSYNCED::
   It is up to you to figure out how to synchronize back to LDAP. This allows {project_name} deployments to support
   updates of user metadata on a read-only LDAP server.  This option only applies when you are importing users from LDAP into the local {project_name} user database.
 
+NOTE: When LDAP provider is created, the set of initial <<_ldap_mappers,LDAP mappers>> is created. The mappers are configured in a "best-effort" way based on
```
```s/configured in a "best-effort"/configured on a "best-effort" basis/g```

mposolda replied:
Changing as you suggested.

iankko commented on server_admin/topics/user-federation/ldap.adoc:
```diff
@@ -51,6 +58,13 @@ UNSYNCED::
   updates of user metadata on a read-only LDAP server.  This option only applies when you are importing users from LDAP into the local {project_name} user database.
 
+NOTE: When LDAP provider is created, the set of initial <<_ldap_mappers,LDAP mappers>> is created. The mappers are configured in a "best-effort" way based on
+      your used LDAP vendor, edit mode and `Import Users` switch. For example in case of UNSYNCED edit mode, the mappers are pre-configured
```
```s/based on your used LDAP vendor, edit mode and `Import Users` switch/based on the chosen combination of the `Vendor`, `Edit Mode`, and `Import Users` switches./g``` 

(or ```based on the combination of ... switches, chosen by you``` if you want)

mposolda replied:
Changing as you suggested.

iankko commented on server_admin/topics/user-federation/ldap.adoc:
```diff
@@ -51,6 +58,13 @@ UNSYNCED::
+      your used LDAP vendor, edit mode and `Import Users` switch. For example in case of UNSYNCED edit mode, the mappers are pre-configured
+      in a way that particular user attribute is preferably read from the DB instead of from the LDAP. However when you later change the edit mode,
+      the mappers configuration will not be changed as it is not easily possible to detect if they were manually changed in the meantime.
+      This means that it is recommended to NOT update edit mode switch, but rather always decide when creating LDAP provider. This applies for
```
s/it is recommended to NOT update edit mode switch/it is recommended NOT to update the `Edit Mode` switch/g

mposolda replied:
Thanks. I've changed this sentence to make it more clear based on your suggestion:

This means that it is recommended NOT to update the `Edit Mode` switch, but rather always decide on `Edit Mode` when creating LDAP provider.

iankko commented on upgrading/topics/keycloak/changes.adoc:
```diff
@@ -1,5 +1,17 @@
+==== LDAP no-import bugfix
+
+In the previous {project_name} version, when LDAP provider was configured with `Import Users` OFF, then it was possible to update the
+user even if some non-LDAP mapped attributes were changed. This may result in confusing behaviour, when the attribute appeared to be updated,
```
s/some/some of/
s/This may result/This resulted/

mposolda replied:
Changing as you suggested

iankko commented on upgrading/topics/keycloak/changes.adoc:
```diff
@@ -1,5 +1,17 @@
+
+In the previous {project_name} version, when LDAP provider was configured with `Import Users` OFF, then it was possible to update the
+user even if some non-LDAP mapped attributes were changed. This may result in confusing behaviour, when the attribute appeared to be updated,
+but it was not. In the current version, the update is not allowed instead of the successful "lost update" .
```
s/, the update is not allowed instead of the successful "lost update"/, the update is not allowed to be performed at all instead to tolerate a successful execution of a "lost update"./g

mposolda replied:
Changing as you suggested. Thanks

iankko commented on upgrading/topics/keycloak/changes.adoc:
```diff
@@ -1,5 +1,17 @@
+but it was not. In the current version, the update is not allowed instead of the successful "lost update" .
+
+This should not affect most of the deployments, but some can be affected under some rare circumstances. For example if you previously
+tried to update user with the admin REST API and the user had some incorrect attribute changes, the update was possible when in the
```
```s/the update was possible when ../the update was possible. With the current version.../g```

or even ```the update was previously possible. With the current version ...``` (if you want)

mposolda replied:
Changing as you suggested.

iankko reviewed (changes requested):
LGTM. Just couple of small nitpicks / text changes to rephrase the sentences a bit.

mposolda commented:
@iankko Thanks, I've changed most of the things in a way you proposed. There is one item to clarify but I believe I've changed it correctly.

iankko reviewed (dismissed):
This version LGTM. Thanks @mposolda !

andymunro commented on server_admin/topics/user-federation/ldap.adoc:
```diff
@@ -17,12 +17,11 @@ Selecting _ldap_ will bring you to the LDAP configuration page.
-also have to be synchronized with your LDAP server as needed.
+The single exception to this is the synchronization of passwords. Passwords are never imported. Their validation is always delegated to the LDAP server.
+The benefits of this approach is that all {project_name} features will work as any extra per-user data that is needed can be stored locally.
+The downside of this approach being each time a specific user is queried for the first time, a corresponding {project_name} database insert is performed.
```
change "being each time" to "is that each time that"

mposolda replied:
Will change as you suggested

andymunro commented on server_admin/topics/user-federation/ldap.adoc:
```diff
@@ -17,12 +17,11 @@ Selecting _ldap_ will bring you to the LDAP configuration page.
+The single exception to this is the synchronization of passwords. Passwords are never imported. Their validation is always delegated to the LDAP server.
+The benefits of this approach is that all {project_name} features will work as any extra per-user data that is needed can be stored locally.
+The downside of this approach being each time a specific user is queried for the first time, a corresponding {project_name} database insert is performed.
+The import may also have to be synchronized with your LDAP server as needed. However, import synchronization is not necessary in
```
Delete "as needed" since you already said it "may" be needed.

mposolda replied:
Will change as you suggested.

andymunro commented on server_admin/topics/user-federation/ldap.adoc:
```diff
@@ -17,12 +17,11 @@ Selecting _ldap_ will bring you to the LDAP configuration page.
+The benefits of this approach is that all {project_name} features will work as any extra per-user data that is needed can be stored locally.
+The downside of this approach being each time a specific user is queried for the first time, a corresponding {project_name} database insert is performed.
+The import may also have to be synchronized with your LDAP server as needed. However, import synchronization is not necessary in
+the case the LDAP mappers are configured to always read particular attributes from the LDAP rather than from the DB.
```
change "the case" to "the case that"

Change DB to database

mposolda replied:
Will change both things you suggest.

andymunro commented on server_admin/topics/user-federation/ldap.adoc:
```diff
@@ -32,6 +31,13 @@ The benefit to this approach is that you do not have the overhead of importing a
 This storage mode is controled by the `Import Users` switch.  Set to `On` to import users.
 
+NOTE: In the case that import of users is disabled, there is no support to save user profile attributes and most of other user metadata
+      into the {project_name} database. The single exception to this are user profile metadata, which are mapped to the LDAP. This possibly includes
```
Try this:
NOTE: If user import is disabled, you cannot save user profile attributes into the {project_name} database. Also you cannot save metadata except for user profile metadata that are mapped to the LDAP.

mposolda replied:
I will change this exactly as you suggested.

andymunro commented on server_admin/topics/user-federation/ldap.adoc:
```diff
@@ -51,6 +57,13 @@ UNSYNCED::
   It is up to you to figure out how to synchronize back to LDAP. This allows {project_name} deployments to support
   updates of user metadata on a read-only LDAP server.  This option only applies when you are importing users from LDAP into the local {project_name} user database.
 
+NOTE: When LDAP provider is created, the set of initial <<_ldap_mappers,LDAP mappers>> is created. The mappers are configured on a "best-effort" basis
```
Add "the" before LDAP

mposolda replied:
Will change as you suggested.

andymunro commented on server_admin/topics/user-federation/ldap.adoc:
```diff
@@ -51,6 +57,13 @@ UNSYNCED::
 
+NOTE: When LDAP provider is created, the set of initial <<_ldap_mappers,LDAP mappers>> is created. The mappers are configured on a "best-effort" basis
+      based on the chosen combination of the `Vendor`, `Edit Mode`, and `Import Users` switches. For example in case of UNSYNCED edit mode, the mappers are pre-configured
+      in a way that particular user attribute is preferably read from the DB instead of from the LDAP. However when you later change the edit mode,
```
add "a" before "particular"
change DB to database

mposolda replied:
Will change as you suggested.

andymunro commented on server_admin/topics/user-federation/ldap.adoc:
```diff
@@ -51,6 +57,13 @@ UNSYNCED::
+      in a way that particular user attribute is preferably read from the DB instead of from the LDAP. However when you later change the edit mode,
+      the mappers configuration will not be changed as it is not easily possible to detect if they were manually changed in the meantime.
+      This means that it is recommended NOT to update the `Edit Mode` switch, but rather always decide on `Edit Mode` when creating
+      LDAP provider. This applies for `Import Users` switch as well.
```
Add "the" before LDAP

mposolda replied:
Changed as you suggested.

andymunro commented on upgrading/topics/keycloak/changes.adoc:
```diff
@@ -1,5 +1,18 @@
+
+==== LDAP no-import bugfix
+
+In the previous {project_name} version, when LDAP provider was configured with `Import Users` OFF, then it was possible to update the
```
Add "the" before LDAP

Remove the "then"  It adds no meaning.

mposolda replied:
Will change both things as you suggested.

andymunro commented on upgrading/topics/keycloak/changes.adoc:
```diff
@@ -1,5 +1,18 @@
+==== LDAP no-import bugfix
+
+In the previous {project_name} version, when LDAP provider was configured with `Import Users` OFF, then it was possible to update the
+user even if some of non-LDAP mapped attributes were changed. This resulted in confusing behaviour, when the attribute appeared to be updated,
```
Also follow "this" noun.   "This situation"
US spelling for words ending in "our" is "or" --- "behavior"

mposolda replied:
Changed both things as you suggested.

andymunro commented on upgrading/topics/keycloak/changes.adoc:
```diff
@@ -1,5 +1,18 @@
+
+In the previous {project_name} version, when LDAP provider was configured with `Import Users` OFF, then it was possible to update the
+user even if some of non-LDAP mapped attributes were changed. This resulted in confusing behaviour, when the attribute appeared to be updated,
+but it was not. In the current version, the update is not allowed to be performed at all instead to tolerate a successful execution
```
Not sure what this means "instead to tolerate a successful execution"  Is it needed?
Are you saying that "it will not tolerate a successful execution?"

iankko replied:
@andymunro For the context, previous behaviour reported success when the expected field / setting wasn't really updated (that's where the name of "lost update" comes from). With the fix, the update is not performed (and the user is informed back, why it wasn't possible to perform it).

The "instead to tolerate a successful execution" is not needed. Maybe better to rephrase the particular sentence as a whole, so it makes more sense.

andymunro replied:
Okay, that sounds like the right idea.  Thanks, @iankko.

mposolda replied:
I've changed this sentence to "In the current version, the update is not allowed to be performed at all if any non-LDAP mapped attributes are changed." Does it sound ok?

iankko replied:
LGTM AFAICT, thanks Marek!

andymunro commented on upgrading/topics/keycloak/changes.adoc:
```diff
@@ -1,5 +1,18 @@
+of a "lost update".
+
+This should not affect most of the deployments, but some can be affected under some rare circumstances. For example if you previously
+tried to update user with the admin REST API and the user had some incorrect attribute changes, the update was possible. With the
```
Add "a" before "user"

mposolda replied:
Will change as you suggested

andymunro reviewed (changes requested):
Looks good, just a few changes to suggest.

mposolda commented:
@andymunro Thanks! I've changed everything as you suggested. I left just one comment unresolved where I am not 100% sure if the new sentence I changed is correct. 

Do you please have a chance to re-review?
